boards/x86: up_squared: enable GPIO

dcpleung opened the pull request:
This PR adds a new GPIO driver for the Intel Apollo Lake SoC, and enables it the UP Squared board configuration. This also includes a sample app to utilize GPIOs on the UP Squared board.

galak commented on dts/bindings/gpio/intel,apl-gpio.yaml:
```diff
@@ -0,0 +1,44 @@
+      category: required
+      description: compatible strings
+      constraint: "intel,apl-gpio"
+
```
`generation: define` is missing

galak commented on boards/x86/up_squared/up_squared.dts:
```diff
@@ -155,5 +155,21 @@
 			status = "ok";
 		};
+
+		gpio: gpio@0 {
```
Why is this in the board.dts vs apollo_lake.dtsi

galak reviewed (dismissed):
Some dt changes needed.

dcpleung committed:
boards/x86: up_squared: updates SYS_CLOCK_HW_CYCLES_PER_SEC

Sets the HPET timer frequency (enabled by default) to 19.2MHz.
And also sets the local APIC timer frequencies according to
the SoC variant.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
gpio: enable callback to specify pin in addition to pin_mask

There are GPIO controllers that can control more than 32 pins.
The pin_mask is no longer adequate in this case. This wraps
the pin_mask in a union together with a field named 'pin'.
The driver is responsible for choosing whether to use
pin_mask or pin.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
boards/x86: up_squared: add board option for SoC variants

There are currently three SoC variants for the UP Squared board:
Pentium, Celeron and Atom. This adds a board option to specify
for which variant is being built.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
boards/x86: up_squared: add sample app for GPIO

This adds a sample app to utilize GPIOs on the UP Squared
board.

Origin: Original

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
gpio: add driver for Intel Apollo Lake SoC

This adds a driver for GPIO controller on the Intel
Apollo Lake SoC.

Origin: Original

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
boards/x86: up_squared: add GPIO dts configuration

This adds the necessary dts entires to enable GPIO controller
on the UP Squared board.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dbkinder reviewed (approved):
works for me, thanks.

galak reviewed (approved):
DT parts look fine to me.
